강원대 BE_이강욱_3주차 과제(1단계)

kanguk01 opened the pull request:
# 과제를 진행하며 생긴 의문점

### 테스트 관련 (+의문점)

단위 테스트는 빠르게!! 라고 생각해서 `Mock` 을 사용하였습니다. 근데 저번 피드백에서 Mock을 지양하라고 하셨는데요, 그래서 이번엔 실제 구현체를 이용하여 테스트를 작성해보았는데 이런 방향이 맞는지 궁금합니다..! 

책을 추천해주시기는 했는데 아직 그걸 읽을 시간은 없어서 짧게 찾아보고 고른 방법인데.. 제가 잘못 접근한것일까요??

### getter관련 의문점

그리고 지금까지는 제가 굳이 getter를 안쓰는 방향으로 진행했는데, 지금 그냥 로직에서는 getter를 안쓰고 잘 구현을 한 것 같은데 테스트코드를 작성하기 위해 `assertEquals(35000, result.getPrice());` 등과 같이 체크하는 과정에서 get을 쓰든 객체 값에 접근을 하든 무조건 이런 방법이 쓰여야되어 고작 테스트때문에 , 테스트코드때문에 그런 메소드를 추가하는 것은 좀 이상한 것 같다고 생각해서 막힌 상태입니다… 이부분에서 너무 고민하느라 시간을 많이 써서 일단은 get을 사용한채로 제출을 드리려고 합니다. (그래서 전체적으로 getter를 사용하였습니다..)

만약 고쳐야한다면 어떤 방식으로 설계를 해야 할까요?? 무조건 get을 사용하면 안된다! 는 아니지만 안쓸 수 있는데 쓰는것은 아니라고 생각을 해서 질문드립니다!

### 엔티티 유효성 관련 의문점

제가 이전엔 

```java
private void validateImageUrl(String imageUrl) {
        if (imageUrl == null || imageUrl.trim().isEmpty()) {
            throw new BusinessException(ErrorCode.INVALID_IMAGE_URL);
        }
    }
```

이런 식의 메소드를 엔티티에 넣어서 객체가 생성 혹은 업데이트 되기 전에 유효성 검증을 엔티티단에서 진행해주었는데, 이제는 `@NotBlank` , `@Min` 등의 유효성을 어노테이션으로 검증하는 것 같습니다. 그래서 어떻게 처리해야할 지 고민중입니다. 

어노테이션을 통한 유효성 검증에서는 `ConstraintViolationException` 를 반환하는 것 같은데, 그렇게되면 제가 이전에 만들었던 `BusinessException` 의 `INVALID_IMAGE_URL` 이런 것들은 아에 필요가 없는 것인지, 아니면 그냥 그대로 사용해도 되는지 궁금합니다. 

일단은 에러 메시지 등의 통일을 위해 그대로 두었는데, 어떤 느낌으로 하는 것이 좋을지 알려주시면 감사하겠습니다!

### 레포지토리 테스트 의문점

```java
package gift.repository;

import gift.entity.Product;
import gift.entity.ProductName;
import gift.entity.Wish;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;

import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
public class WishRepositoryTest {

    @Autowired
    private WishRepository wishRepository;

    @Autowired
    private ProductRepository productRepository;

    @Test
    public void 위시리스트_저장_조회() {
        Product product = new Product(new ProductName("오둥이 입니다만"), 29800, "https://example.com/product1.jpg");
        productRepository.save(product);

        Wish wish = new Wish(1L, product.getId());
        wishRepository.save(wish);

        List<Wish> wishes = wishRepository.findByUserId(1L);
        assertEquals(1, wishes.size());
        assertEquals(product.getId(), wishes.get(0).getProductId());
    }

    @Test
    public void 위시리스트_삭제() {
        Product product = new Product(new ProductName("오둥이 입니다만"), 29800, "https://example.com/product1.jpg");
        productRepository.save(product);

        Wish wish = new Wish(1L, product.getId());
        wishRepository.save(wish);

        wishRepository.delete(wish);

        List<Wish> wishes = wishRepository.findByUserId(1L);
        assertTrue(wishes.isEmpty());
    }
}

```

이렇게 WishRepository를 적었는데 얘만 실패합니다. 

`Product` 테이블이 생성되지 않는다고 하는데.. 실제로 실행하면 잘 되는데  테스트코드에서만 안됩니다

`spring.jpa.hibernate.ddl-auto=update`  를 설정에 추가해 보았는데도 안되는데, 혹시 테스트에서 다른 객체의 테이블을 생성하기 위해 별도의 작업을 해야 하는 건가요? 

아니면 schema.sql을 따로 작성해서 테스트 전에 주입시켜줘야하는건가요?

### 마지막으로
현재 엔티티 설계가 좀 이상한 듯 싶은데 의도된 과제 흐름인 것 같아서 step2에 수정할 예정입니다.

그리고 전체적인 코드 흐름과 구조 설계 등이 괜찮게 흘러가고있는지 궁금합니다.. 저번에 느낀 것인데 피드백을 수정하다보니 의도치않게 바뀌는부분도 생기고 분명 생각을 하면서 짰는데 어느새 보면 이상하게 돼있고.. 그런 부분이 보여서 약간 신경이 쓰이네요 🥲 신경써서 열심히 하겠습니다..

항상 시간내주셔서 피드백 주셔서 감사합니다. 늘 고생 많으십니다 멘토님!!🥹

kanguk01 committed:
docs : README.md 작성

docs : README.md 작성

과제 요구사항에 따라 리드미를 작성한다. JPA적용, 테스트코드 작성 내용을 포함한다.

kanguk01 committed:
set : build.gradle, application.properties 설정

set : build.gradle, application.properties 설정

JPA 추가에 따라 의존성 추가 및 설정을 추가한다.

kanguk01 committed:
refactor : Product JPA사용 리팩토링

refactor : Product JPA사용 리팩토링

Product 객체에 관련된 파일들을 JPA를 사용하도록 리팩토링한다.

kanguk01 committed:
test : ProductService 단위테스트 작성

test : ProductService 단위테스트 작성

mock 대신 실제 객체를 이용하도록 단위테스트를 작성한다.

kanguk01 committed:
refactor : User JPA사용 리팩토링

refactor : User JPA사용 리팩토링

User 객체에 관련된 파일들을  JPA를 사용하도록 리팩토링한다.

kanguk01 committed:
test : UserService 단위테스트 작성

test : UserService 단위테스트 작성

UserService 단위테스트를 작성한다.

kanguk01 committed:
refactor : Wish JPA사용 리팩토링

refactor : Wish JPA사용 리팩토링

Wish 객체에 관련된 파일들을 JPA를 사용하도록 리팩토링한다.

kanguk01 committed:
test : WishService 단위테스트 작성

test : WishService 단위테스트 작성

WishService 단위테스트를 작성한다.

kanguk01 committed:
test : 레포지토리 테스트 작성

test : 레포지토리 테스트 작성

@DataJpaTest 이용하여 레포지토리 테스트를 작성한다.

kanguk01 committed:
refactor : user 테이블 이름 변경

refactor : user 테이블 이름 변경

의도치 않게 users로 바뀐 테이블 이름을 다시 user로 수정한다. 예약어를 피하기 위해 `를 사용한다.

pci2676 commented on src/test/java/gift/service/UserServiceTest.java:
```diff
@@ -1,156 +1,100 @@
-import static org.mockito.Mockito.*;
 
+@SpringBootTest
+@Transactional
```
테스트 클래스에는 transactional을 붙이지 않는것이 좋다고 생각하는데요.

혹시 어떠한 이유에서 사용하셨을까요?

아래 글은 제가 예전에 작성한 글인데 참고해보시면 좋을것 같습니다. :)  
https://javabom.tistory.com/103

kanguk01 replied:
헉 단지 롤백을 위해 사용했던 것인데, 블로그 글 읽어보았습니다. 수정하도록 하겠습니다 감사합니다!

pci2676 commented:
> ### 엔티티 유효성 관련 의문점
> 제가 이전엔
> 
> ```java
> private void validateImageUrl(String imageUrl) {
>         if (imageUrl == null || imageUrl.trim().isEmpty()) {
>             throw new BusinessException(ErrorCode.INVALID_IMAGE_URL);
>         }
>     }
> ```
> 
> 이런 식의 메소드를 엔티티에 넣어서 객체가 생성 혹은 업데이트 되기 전에 유효성 검증을 엔티티단에서 진행해주었는데, 이제는 `@NotBlank` , `@Min` 등의 유효성을 어노테이션으로 검증하는 것 같습니다. 그래서 어떻게 처리해야할 지 고민중입니다.
> 
> 어노테이션을 통한 유효성 검증에서는 `ConstraintViolationException` 를 반환하는 것 같은데, 그렇게되면 제가 이전에 만들었던 `BusinessException` 의 `INVALID_IMAGE_URL` 이런 것들은 아에 필요가 없는 것인지, 아니면 그냥 그대로 사용해도 되는지 궁금합니다.
> 
> 일단은 에러 메시지 등의 통일을 위해 그대로 두었는데, 어떤 느낌으로 하는 것이 좋을지 알려주시면 감사하겠습니다!

검증용 애너테이션은 그 설계가 클라이언트로부터 입력받을때 사용하라고 만들어진것으로 알고 있기때문에

저는 entity에서는 validate를 할 수 있는 애너테이션을 활용하는 편은 아닙니다.

validate를 할 수 있는 애너테이션은 dto 레벨에서 사용하고 entity에서는 도메인 로직으로 validate 항목을 표현하는편입니다.

그러면 dto에서 검증하고 entity에서 도메인 검증을 하며 검증 내역이 중복되는 상황이 벌어질 수 있는데

저는 dto에서 검증하는 내역을 최소한의 검증만 하도록 합니다. (null인지 빈 문자열인지 등)

pci2676 commented:
> ### getter관련 의문점
> 그리고 지금까지는 제가 굳이 getter를 안쓰는 방향으로 진행했는데, 지금 그냥 로직에서는 getter를 안쓰고 잘 구현을 한 것 같은데 테스트코드를 작성하기 위해 `assertEquals(35000, result.getPrice());` 등과 같이 체크하는 과정에서 get을 쓰든 객체 값에 접근을 하든 무조건 이런 방법이 쓰여야되어 고작 테스트때문에 , 테스트코드때문에 그런 메소드를 추가하는 것은 좀 이상한 것 같다고 생각해서 막힌 상태입니다… 이부분에서 너무 고민하느라 시간을 많이 써서 일단은 get을 사용한채로 제출을 드리려고 합니다. (그래서 전체적으로 getter를 사용하였습니다..)
> 
> 만약 고쳐야한다면 어떤 방식으로 설계를 해야 할까요?? 무조건 get을 사용하면 안된다! 는 아니지만 안쓸 수 있는데 쓰는것은 아니라고 생각을 해서 질문드립니다!

오 정확히 어느 부분인지 잘 모르겠습니다. entity는 mapper때문에라도 getter가 사용되어야 하는 상황같아보여요~

이렇게 궁금한 포인트가 있으면 PR 생성후 commet기능을 이용해 code에 남겨주시면 도움이 될 것 같습니다. 😄

pci2676 commented:
> ### 레포지토리 테스트 의문점
> ```java
> package gift.repository;
> 
> import gift.entity.Product;
> import gift.entity.ProductName;
> import gift.entity.Wish;
> import org.junit.jupiter.api.Test;
> import org.springframework.beans.factory.annotation.Autowired;
> import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
> 
> import java.util.List;
> import java.util.Optional;
> 
> import static org.junit.jupiter.api.Assertions.*;
> 
> @DataJpaTest
> public class WishRepositoryTest {
> 
>     @Autowired
>     private WishRepository wishRepository;
> 
>     @Autowired
>     private ProductRepository productRepository;
> 
>     @Test
>     public void 위시리스트_저장_조회() {
>         Product product = new Product(new ProductName("오둥이 입니다만"), 29800, "https://example.com/product1.jpg");
>         productRepository.save(product);
> 
>         Wish wish = new Wish(1L, product.getId());
>         wishRepository.save(wish);
> 
>         List<Wish> wishes = wishRepository.findByUserId(1L);
>         assertEquals(1, wishes.size());
>         assertEquals(product.getId(), wishes.get(0).getProductId());
>     }
> 
>     @Test
>     public void 위시리스트_삭제() {
>         Product product = new Product(new ProductName("오둥이 입니다만"), 29800, "https://example.com/product1.jpg");
>         productRepository.save(product);
> 
>         Wish wish = new Wish(1L, product.getId());
>         wishRepository.save(wish);
> 
>         wishRepository.delete(wish);
> 
>         List<Wish> wishes = wishRepository.findByUserId(1L);
>         assertTrue(wishes.isEmpty());
>     }
> }
> ```
> 
> 이렇게 WishRepository를 적었는데 얘만 실패합니다.
> 
> `Product` 테이블이 생성되지 않는다고 하는데.. 실제로 실행하면 잘 되는데 테스트코드에서만 안됩니다
> 
> `spring.jpa.hibernate.ddl-auto=update` 를 설정에 추가해 보았는데도 안되는데, 혹시 테스트에서 다른 객체의 테이블을 생성하기 위해 별도의 작업을 해야 하는 건가요?
> 
> 아니면 schema.sql을 따로 작성해서 테스트 전에 주입시켜줘야하는건가요?


test 패키지에 resources 폴더에도 application.properties를 생성해주세요.

내부가 비어있는 application.properties를 만들어두면 main 에 위치한 application.properties의 속성을 그대로 가져다 사용하게 됩니다.

제가 보았을때 이렇게 조치하면 문제가 생기지 않을것 같아보여요

pci2676 commented:
> ### 마지막으로
> 현재 엔티티 설계가 좀 이상한 듯 싶은데 의도된 과제 흐름인 것 같아서 step2에 수정할 예정입니다.
> 
> 그리고 전체적인 코드 흐름과 구조 설계 등이 괜찮게 흘러가고있는지 궁금합니다.. 저번에 느낀 것인데 피드백을 수정하다보니 의도치않게 바뀌는부분도 생기고 분명 생각을 하면서 짰는데 어느새 보면 이상하게 돼있고.. 그런 부분이 보여서 약간 신경이 쓰이네요 🥲 신경써서 열심히 하겠습니다..
> 
> 항상 시간내주셔서 피드백 주셔서 감사합니다. 늘 고생 많으십니다 멘토님!!🥹

이러한 부분도 이상하다고 생각되는 부분은 commet로 남겨주시면 제가 확인하기 좋을것 같아요~

pci2676 commented on src/test/java/gift/service/UserServiceTest.java:
```diff
@@ -1,156 +1,100 @@
     @BeforeEach
     public void setUp() {
-        MockitoAnnotations.openMocks(this);
+        userService = new UserService(userRepository, tokenService);
```
userService를 한번에 spring bean으로 주입해서 사용해도 되지 않나요?

kanguk01 replied:
오 맞습니다! 효율적인 코드를 위해 더 노력해보겠습니다 감사합니다!!

pci2676 commented on src/main/java/gift/service/WishService.java:
```diff
@@ -1,60 +1,55 @@
-        Wish createdWish = wishDao.insertWish(wish);
+        validateProductId(wishRequestDto.getProductId());
+        Wish wish = new Wish(userId, wishRequestDto.getProductId());
+        Wish createdWish = wishRepository.save(wish);
 
-        ProductResponseDto product = productService.getProductById(wishRequestDto.productId);
+        ProductResponseDto product = productService.getProductById(wishRequestDto.getProductId());
 
-        return toWishResponseDto(createdWish, product);
+        return WishMapper.toWishResponseDto(createdWish, product);
```
처음에 product를 찾아와 wish를 만들때 사용하고 이후에 추가 조회 없이 반환값을 만들때 사용하면

product의 조회를 1회로 줄일 수 있을것 같습니다.

kanguk01 replied:
거슬렸던 부분이기는 했는데 그 방향으로 수정해보도록 하겠습니다!

pci2676 reviewed (changes requested):
질문에 대한 답변과 리뷰를 조금 남겨두었으니 확인해주세요.

kanguk01 committed:
test : WishRepositoryTest 추가

test : WishRepositoryTest 추가

기존 테스트가 정상적으로 작동되지 않던 문제를 해결하여 WishRepositoryTest를 추가한다.

kanguk01 committed:
refactor : 엔티티의 유효성 검증 로직 수정

refactor : 엔티티의 유효성 검증 로직 수정

Dto에서는 어노테이션을 이용한 유효성 검증, 엔티티에서는 도메인 로직으로 검증하도록 수정한다.

kanguk01 committed:
refactor : 테스트코드에서 스프링 빈으로 의존성 주입

refactor : 테스트코드에서 스프링 빈으로 의존성 주입

테스트 코드에서 굳이 생성자주입 하는 대신 한번에 spring bean으로 주입하도록 한다.

kanguk01 committed:
refactor : 서비스 레이어 테스트 @Transactional 삭제

refactor : 서비스 레이어 테스트 @Transactional 삭제

위험성이 있는 @Transactional 를 삭제하여 @AfterEach로 값을 되돌리도록 리팩토링한다.

kanguk01 committed:
refactor : 코드 최적화하여 DB접근 최소화

refactor : 코드 최적화하여 DB접근 최소화

필요없이 두번씩 DB에 접근하던 코드들을 최적화하여 접근 횟수를 줄인다.

kanguk01 commented:
피드백주신 부분 수정하였습니다! 피드백해주실때 걸어주신 코드 외에도 수정 사항 있으면 같이 수정해보았는데 확인해주시면 감사하겠습니다!

그리고 

>단위 테스트는 빠르게!! 라고 생각해서 Mock 을 사용하였습니다. 근데 저번 피드백에서 Mock을 지양하라고 하셨는데요, 그래서 이번엔 실제 구현체를 이용하여 테스트를 작성해보았는데 이런 방향이 맞는지 궁금합니다..!
책을 추천해주시기는 했는데 아직 그걸 읽을 시간은 없어서 짧게 찾아보고 고른 방법인데.. 제가 잘못 접근한것일까요??
>

이 부분에 관하여 답변을 안해주셨는데, 이 방향이 맞는 것인가요?!

pci2676 commented:
> 피드백주신 부분 수정하였습니다! 피드백해주실때 걸어주신 코드 외에도 수정 사항 있으면 같이 수정해보았는데 확인해주시면 감사하겠습니다!
> 
> 그리고
> 
> > 단위 테스트는 빠르게!! 라고 생각해서 Mock 을 사용하였습니다. 근데 저번 피드백에서 Mock을 지양하라고 하셨는데요, 그래서 이번엔 실제 구현체를 이용하여 테스트를 작성해보았는데 이런 방향이 맞는지 궁금합니다..!
> > 책을 추천해주시기는 했는데 아직 그걸 읽을 시간은 없어서 짧게 찾아보고 고른 방법인데.. 제가 잘못 접근한것일까요??
> 
> 이 부분에 관하여 답변을 안해주셨는데, 이 방향이 맞는 것인가요?!

네 지금처럼 작성해주시면 좋을것 같습니다. 😄 

그런데 mock을 사용하는게 무조건 좋지 않은것은 아니니 상황에 따라 사용해야한다는 점을 기억해주세요~

pci2676 commented on src/main/java/gift/service/WishService.java:
```diff
@@ -1,60 +1,54 @@
         return wishes.stream()
-                .map(this::toWishResponseDto)
+                .map(wish -> {
+                    ProductResponseDto product = productService.getProductById(wish.getProductId());
```
오 이러면 for-loop에서 조회하게 되면서 조회가 여러건 발생하게 될 것 같습니다.

이렇게 되면 성능상 이슈가 발생할 수 있어 in절을 이용한 조회후 dto를 만드시는게 더 유리할 것 같습니다.

kanguk01 replied:
오 한번 체크해보겠습니다

pci2676 reviewed (approved):
전체적으로 깔끔하게 작성해주신 것 같습니다.

merge하겠습니다. 🚀
